cpp/unreal_plugins/SimulationController/Source/SimulationController/UrdfBotAgent.cpp

```diff
@@ -0,0 +1,142 @@
+
+#include <Components/SceneCaptureComponent2D.h>
+
+#include "CoreUtils/Assert.h"
+#include "CoreUtils/Box.h"
+#include "CoreUtils/Config.h"
+#include "CoreUtils/Std.h"
+#include "CoreUtils/Unreal.h"
+#include "SimulationController/CameraSensor.h"
+#include "SimulationController/ImuSensor.h"
+#include "SimulationController/SonarSensor.h"
+#include "UrdfBot/UrdfBotPawn.h"
+#include "UrdfBot/UrdfLinkComponent.h"
+#include "UrdfBot/UrdfRobotComponent.h"
+#include "UrdfBot/UrdfParser.h"
+
+UrdfBotAgent::UrdfBotAgent(UWorld* world)
+{
+    FActorSpawnParameters actor_spawn_params;
+    actor_spawn_params.Name = Unreal::toFName(Config::get<std::string>("SIMULATION_CONTROLLER.URDFBOT_AGENT.URDFBOT_ACTOR_NAME"));
+    actor_spawn_params.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
+    urdf_bot_pawn_ = world->SpawnActor<AUrdfBotPawn>(FVector::ZeroVector, FRotator::ZeroRotator, actor_spawn_params);
+    ASSERT(urdf_bot_pawn_);
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "camera")) {
+        camera_sensor_ = std::make_unique<CameraSensor>(
+            urdf_bot_pawn_->camera_component_,
+            Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.CAMERA.RENDER_PASSES"),
+            Config::get<unsigned int>("SIMULATION_CONTROLLER.URDFBOT_AGENT.CAMERA.IMAGE_WIDTH"),
+            Config::get<unsigned int>("SIMULATION_CONTROLLER.URDFBOT_AGENT.CAMERA.IMAGE_HEIGHT"));
+        ASSERT(camera_sensor_);
+
+        // update FOV
+        for (auto& pass : Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.CAMERA.RENDER_PASSES")) {
+            camera_sensor_->render_passes_.at(pass).scene_capture_component_->FOVAngle =
+                Config::get<float>("SIMULATION_CONTROLLER.URDFBOT_AGENT.CAMERA.FOV");
+        }
+    }
+}
+
+UrdfBotAgent::~UrdfBotAgent()
+{
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "camera")) {
+        ASSERT(camera_sensor_);
+        camera_sensor_ = nullptr;
+    }
+
+    ASSERT(urdf_bot_pawn_);
+    urdf_bot_pawn_->Destroy();
+    urdf_bot_pawn_ = nullptr;
+}
+
+void UrdfBotAgent::findObjectReferences(UWorld* world) {}
+
+void UrdfBotAgent::cleanUpObjectReferences() {}
+
+std::map<std::string, Box> UrdfBotAgent::getActionSpace() const
+{
+    std::map<std::string, Box> action_space;
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.ACTION_COMPONENTS");
+
+    std::map<std::string, Box> robot_component_action_space = urdf_bot_pawn_->urdf_robot_component_->getActionSpace(action_components);
+    for (auto&  robot_component_observation_action_component : robot_component_action_space) {
```

Comment:
`for (auto& robot_component_observation_action_component : robot_component_action_space) {`

(removed extra space)